anvil-constraintlayout/src/main/kotlin/trikita/anvil/ConstraintDsl.kt

```diff
@@ -0,0 +1,579 @@
+@file:Suppress("DEPRECATION", "UNCHECKED_CAST", "MemberVisibilityCanBePrivate", "unused")
+
+package trikita.anvil
+
+import android.view.View
+import androidx.constraintlayout.widget.*
+import java.util.*
+
+// TODO many of these methods should be available for ConstraintLayout but not ConstraintLayout itself.
+// Revisit this when multi-receivers are available.
+
+fun constraintLayout(configure: ConstraintLayoutScope.() -> Unit = {}) =
+    v<ConstraintLayout>(configure.bind(ConstraintLayoutScope))
+abstract class ConstraintLayoutScope : ViewGroupScope() {
+    fun type(arg: Int): Unit = attr("type", arg)
+    fun referencedIds(vararg arg: Int): Unit = attr("referencedIds", arg)
+    fun leftConstraintToLeft(toId: Int): Unit = attr("leftConstraintToLeft", toId)
+    fun leftConstraintToRight(toId: Int): Unit = attr("leftConstraintToRight", toId)
+    fun leftConstraintToTop(toId: Int): Unit = attr("leftConstraintToTop", toId)
+    fun leftConstraintToBottom(toId: Int): Unit = attr("leftConstraintToBottom", toId)
+    fun leftConstraintToParent(): Unit = leftConstraintToLeft(ConstraintSet.PARENT_ID)
+    fun rightConstraintToRight(toId: Int): Unit = attr("rightConstraintToRight", toId)
+    fun rightConstraintToLeft(toId: Int): Unit = attr("rightConstraintToLeft", toId)
+    fun rightConstraintToTop(toId: Int): Unit = attr("rightConstraintToTop", toId)
+    fun rightConstraintToBottom(toId: Int): Unit = attr("rightConstraintToBottom", toId)
+    fun rightConstraintToParent(): Unit = rightConstraintToRight(ConstraintSet.PARENT_ID)
+    fun topConstraintToRight(toId: Int): Unit = attr("topConstraintToRight", toId)
+    fun topConstraintToLeft(toId: Int): Unit = attr("topConstraintToLeft", toId)
+    fun topConstraintToTop(toId: Int): Unit = attr("topConstraintToTop", toId)
+    fun topConstraintToBottom(toId: Int): Unit = attr("topConstraintToBottom", toId)
+    fun topConstraintToParent(): Unit = topConstraintToTop(ConstraintSet.PARENT_ID)
+    fun bottomConstraintToRight(toId: Int): Unit = attr("bottomConstraintToRight", toId)
+    fun bottomConstraintToLeft(toId: Int): Unit = attr("bottomConstraintToLeft", toId)
+    fun bottomConstraintToTop(toId: Int): Unit = attr("bottomConstraintToTop", toId)
+    fun bottomConstraintToBottom(toId: Int): Unit = attr("bottomConstraintToBottom", toId)
+    fun bottomConstraintToParent(): Unit = bottomConstraintToBottom(ConstraintSet.PARENT_ID)
+    fun chain(
+        leftId: Int,
+        leftSide: Int,
+        rightId: Int,
+        rightSide: Int,
+        chainIds: IntArray?,
+        weights: FloatArray?,
+        style: Int
+    ): Unit = attr(
+        "chain",
+        ConstraintChain(leftId, leftSide, rightId, rightSide, chainIds, weights, style))
+    fun circleConstraint(id: Int, angle: Int, radius: Int): Unit =
+        attr("circleConstraint", CircleConstraint(0, id, angle, radius))
+
+    // TODO was id, should override?
```

Comment:
No, CL handles id with its own way

Reply:
I've tried to do that, but overriding is not working for DSLs the way one would expect: when you call `id(...)` on child of coordinator, you still call e.g. `TextViewScope.id(...)`. So the only way to have this working properly is somehow implement behavior variance in setter: if name is id and view has CoordinatorLayout as direct parent, then run coordinator's variant of setter, not View's one. Which is kind of hard as core setter should somehow know there is an override for it. So current dumb sequential run on all setters with switching on attribute name in each one won't work.

I guess we'll have to live with current situation for now, having unique property and a huge warning that a) you have to use `constraintId` for CL children and b) this won't work as expected for composite views which use plain `id` property. I'll file separate issue on upgrading this behavior tomorrow.

Reply:
Filed #27 on this.